Add support for GDT40 (Gaussian grid).

ajdawson opened the pull request:
Support for loading messages on regular Gaussian grids in the new GRIB2 loading framework.

marqh commented:
The GRIB2 grid definition template 3.40 defines a different number of x/i points for each y/j parallel in the horizontal plane.  The data payload in the message is not a 2D array of data.

I think this pull request creates a cube with a regular horizontal grid in latitude and longitude, interpreted from the gaussian grid, which would mean that the data would be resampled during ingestion; is this the case?
If so, I missed the code which does this in my review.

Recent contact with modellers using this grid suggests that a CF representation which preserves the gaussian nature of the model grid is strongly preferred by them.

This could be delivered by producing a Cube with a 1D data array, a 1D latitude and a 1D longitude coordinate, with values sourced explicitly from the GRIB message.

The resampling approach I think you are using here could then be added as an optional post load step, wrapped around a gaussian aware load operation?

Would this approach meet your requirements?

ajdawson commented:
This is complex and I just wrote a whole long reply to you before I realised we have crossed wires at both ends!

My intent was to implement load support for *regular* Gaussian grids, which are defined by GDT40. These grids are rectangular and simply have unevenly spaced latitudes, so no resampling is necessary at all. However, I now realised that the reduced/quasi-Gaussian, which is what you are describing, is also implemented by GDT40.

This implementation simply loads what is in the file, no resampling is performed. Therefore I think the simplest route to take is to add a check for either `Ni` or `Nj` being missing (the indicator of a reduced grid) and raise an exception for an unsupported message. How does that sound?

ajdawson commented:
Although I just tested and it seems there is already code in place at a higher level to detect quasi-regular grids, I get this error when loaded a reduced Gaussian grid:

```
TranslationError: Grid definition section 3 contains unsupported quasi-regular grid
```

marqh commented:
Crossed wires indeed, I had not comprehended that there was 1 grid definition template to support both of these cases and responded in haste.

I now understand the approach.  If the gaussian grid is regular then it is represented as a 2D data array located by 1D latitudes and longitudes, which is what you have implemented; that sounds fine to me.

I did not spot the identification of quasi-regular fields as unsupported in the code.  Indeed the test for this is located elsewhere:
https://github.com/ajdawson/iris/blob/grib2-gdt40/lib/iris/fileformats/grib/_message.py#L98

Although this is not testing for Ni or Nj to be missing.  Instead this is referencing :
> Section 3 - Grid Definition Section
> 11 Number of octets for optional list of numbers (see Note 2)
> 12 Interpretation of list of numbers (see Code table 3.11)
> (2) An optional list of numbers may be used to document a quasi-regular grid. In such a case, octet 11 is non zero and gives the number of octets used per item on the list. For all other cases, such as regular grids, octets 11 and 12 are zero and no list is appended to the grid definition template.

It looks to me like your approach is safe and sensible.  There can be an additional piece of work to implement quasi-regular grids (gaussian or otherwise) independent of this activity.

I'll go back through the code in detail and provide detailed feedback; thank you for clearing up my confusion.

marqh commented on lib/iris/fileformats/grib/_load_convert.py:
```diff
@@ -758,6 +758,59 @@ def check_range(v1, v2, n, d):
+    circular = _is_circular(x_points, 360.0)
+
+    # Get the latitude points.
+    y_points = section['distinctLatitudes']
```
I think that distinctLatitudes is a key computed within the GRIB API.  We are trying to use coded keys which have a 1:1 relationship with an entry in the WMO manual on codes.

Can you replace the use of this computed key with a coded key?

ajdawson replied:
>Can you replace the use of this computed key with a coded key?

It is possible yes. It isn't all that straightforward though, you have to solve for the roots of an appropriate Legendre polynomial, which involves computing eigenvalues of a (possibly large) matrix. I have the code to do this, and we can do some caching to mitigate and performance issues which would help when loading many messages with the same grid.

>We are trying to use coded keys which have a 1:1 relationship with an entry in the WMO manual on codes.

To get me up to speed, why was this path chosen? I'm guessing to be less reliant on ECMWF's implementation?

marqh replied:
that is the thought process, yes. We are still investigating its applicability

rather than provide this in code, could you provide me with a brief overview of how this would be retrieved, in principle, from the GRIB2 encoding?

thank you

ajdawson replied:
The number of latitudes between the equator and pole is enough to define a Gaussian grid, and this is coded in the message (`N`). From this we need to calculate the Gaussian latitudes that correspond to this number.

The calculation just involves defining `2 * N ` Legendre coefficients (trivial, all `0` except the last one which is `1`) and solving for the roots of the polynomial, which is done by first computing the companion matrix and finding the eigenvalues (and a bit of numerical massaging for better estimates of the roots). The roots are directly related to the latitudes by `latitudes = arcsin(roots)`.

The previous step gives us latitudes for the full globe, we'd then have to take a subset based on the encoded first and last latitudes (for a subset of a Gaussian grid) I think, I may be misinterpreting this part.

This would need to be done for every GDT40 message, and with large grid sizes the calculation of latitudes could take a few seconds so caching would probably be important (but easy to do).

marqh commented on lib/iris/fileformats/grib/_message.py:
```diff
@@ -371,7 +371,7 @@ def _get_key_value(self, key):
                        'scaleFactorOfCentralWaveNumber',
                        'scaledValueOfCentralWaveNumber',
-                       'longitudes', 'latitudes')
+                       'longitudes', 'latitudes', 'distinctLatitudes')
```
as above: can we use a coded key here?

ajdawson commented:
As a side note to the discussion about computed vs coded keys:

I see that the code for loading grid definition templates 4 and 5 makes use of the `latitudes` and `longitudes` keys. These must be computed keys right? I'm just wondering if there are shades of grey in terms of using some computed keys when it avoids complex calculations, or if the intent really is to compute everything ourselves from coded values?

rhattersley commented:
> I see that the code for loading grid definition templates 4 and 5 makes use of the latitudes and longitudes keys. These must be computed keys right?

Actually, no. Templates 4 and 5 are variable resolution grids, so the messages contain explicit lists of latitudes and longitudes.

> I'm just wondering if there are shades of grey in terms of using some computed keys

I'm in favour of the ":sunglasses: of grey" approach.

The decision to avoid computed keys had a number of drivers, including:
 * Improving the comprehensibility and transparency of the translation process.
 * Minimising the impact from changes in interpretation applied.
 * Avoiding being wedded to the ECMWF GRIB API.

If we can avoid a ton of work by using a computed key, and the usage of that computed key doesn't undermine the ability to comprehend the translation then it's worthy of consideration. Can we use it in such a way that we could easily replace it with our own implementation using coded keys?

ajdawson commented:
Thanks for the input @rhattersley. In this case we can definitely implement it using our own functionality, but I'd say it would add significant complexity to the code, and in my opinion make it harder to understand. However, I really don't mind which way this goes, I already have the code needed to provide our own generation of Gaussian latitudes, so ease of implementation needn't be a consideration, this is a design issue.

Please advise on what would be the preferred path.

marqh commented:
I wonder if we can tread the shady path here, by including a description, relating back to some 'WMO manual on codes' terms, of our understanding of what 'distinctLatitudes' does.

This seems to be what you have already done in the PR comments @ajdawson: perhaps this could be included as a code comment or function docstring.   This would enable us to capture the intent in our code base, but still use the gribAPI key; kind of use by exception.

This also sits well with your earlier comments about running a calculation for each message with the same expected results and wanting to cache the result (which is going to add more complication if it's put into the GRIB loader).

marqh commented on lib/iris/fileformats/grib/_load_convert.py:
```diff
@@ -758,6 +758,59 @@ def check_range(v1, v2, n, d):
+
+    # Get the latitude points.
+    y_points = section['distinctLatitudes']
+    y_points.sort()
```
is this sort required because of hte way the API returns distinctLatitudes?  It would be good to understand this step (sorry if it's obvious)

ajdawson replied:
It was inherited from the previous implementation. I just did a quick test and it seems this may not be necessary as the `distinctLatitudes` key seems to return values consistent with the `jScansPositively` key. I don't remember how we/I made the original decision to sort and reorder.

rhattersley commented:
> we can definitely implement it using our own functionality, but I'd say it would add significant complexity to the code, and in my opinion make it harder to understand.

:+1: for sticking with the computed key with a comment.

> I wonder if we can tread the shady path here, by including a description, relating back to some 'WMO manual on codes' terms, of our understanding of what 'distinctLatitudes' does.

If it's not already written down somewhere authorative (e.g. not Wikipedia :wink:) perhaps the WMO manual could be extended?

ajdawson commented:
I've added a new commit which adds an explanation of why we use the computed key instead of values derived from coded keys. I also removed the sort-and-reorder as it seems not to be necessary.

marqh commented on lib/iris/fileformats/grib/_load_convert.py:
```diff
@@ -758,6 +758,85 @@ def check_range(v1, v2, n, d):
+    # GRIB2 message. This computed key provides a rapid calculation of the
+    # monotonic latitude points that form the Gaussian grid, accounting for
+    # the direction and coverage of the grid.
+    y_points = section['distinctLatitudes']
```
that is really helpful, thank you

marqh commented:
This pull request looks good to me, I think it can be happily merged.

I think that it is worth considering an integration test for this template, given our dependence on the calculation of distinctlatitudes from the GRIB API.  

if it's easy to extend 
https://github.com/SciTools/iris/blob/master/lib/iris/tests/integration/test_grib2.py
to load a small gdt40 grib2 message and only test that the latitude and longitude coordinates are loaded correctly then I would view that as an benefit.  I'm content to leave this out if it's viewed as unnecessary.

ajdawson commented:
Looks like the sort-and-reorder of latitude values was necessary after all... brb.

ajdawson commented:
I've taken action to correct the direction of the latitude values, it seems like the `distinctLatitudes` key returns values in ascending order (but I can't guarantee that, hence the sort).

I've added an integration test too, which proved its worth already as it allowed me to find the aforementioned bug!

marqh commented:
i think you're just missing a date stamp change on the integration test module

ajdawson commented:
OK, all done @marqh / @rhattersley.

marqh commented:
this all looks good to merge from my point of view

rhattersley commented:
Let's do it!

rhattersley commented:
Thanks chaps!
